### PR TITLE
feat: expand gh-r zunit tests

### PR DESCRIPTION
**Pull request by vladdoster**
## Usage examples

```zsh

```

## Types of changes 

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation change
- [ ] New feature (non-breaking change which adds functionality)

## Checklist: 

- [ ] All new and existing tests passed.
- [ ] I have added tests to cover my changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: tests/_support/gen-ghr-test [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+# Usage:
+#   gen-ghr-test \
+#       'ouch-org/ouch' \
+#       'sharkdp/pastel' \
+#       'Elkowar/pipr' \
+#       'jqnatividad/qsv' \
+#       'lotabout/rargs' \
+#       'nivekuil/rip' \
+#       'sirwart/ripsecrets' \
+#       'ismaelgv/rnr'
```
Comment (vladdoster) [inline tests/_support/gen-ghr-test:14]:
```suggestion
#   gen-ghr-test [<user/repository>...]
```

File: tests/_support/gen-ghr-test [hunk 2]
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env zsh
```
Comment (jankatins) [inline tests/_support/gen-ghr-test:1]: At one point it might result in faster and more stable cycles to refactor the logic for selecting the right file into an extra function and build a corpus of api call result + annotated "wanted file for each OS" and ran that in a unit test. 

And then use this script only as a weekly test run or so :-)

Comment (vladdoster) [reply tests/_support/gen-ghr-test:1]: @jankatins,

Sorry, I don't quite follow. This was a script I created not to have to copy-paste when I wanted to create multiple tests, and it isn't used in any test process. I just found it helpful and added it for others.

Are you suggesting creating mocks? If yes...

IMO, automated testing involves many different techniques, but one that shouldn’t be used is mocking. Mocks are a distraction at best and provide false confidence at worst.

The current test paradigm allows for an actual E2E test of Zinit other than deletion (which we should add) and ensures that annexes are correctly working.

Regarding gh-r tests, 75% is duplicate code and could be encapsulated in a function.

Comment (jankatins) [reply tests/_support/gen-ghr-test:1]: No, not mocks, only testing the "find the right file to download" functionality based on real but saved responses.

From my understanding, the ghr test tests do the following:
- the ghr functionality, but the code path is very narrow despite having a lot of test cases -> Essentially it's a lot of duplicated tests: e.g. if two projects use the same github action to upload releases, there is no need to test both projects.
- The "file to download selection" logic, but in a fragile and slow manner: if a projects removes a file or renames the files, the tests might fail. To run all, you need to download, extract, and run all, which is really slow.

My suggestion was to refactor out the second part and make it narrowly scoped: get all the responses, dump them as examples into a test folder and annotate them: "with OS set to linux, Arch to x86, and the input file a, the filename should be ...". This test should essentially finish in a second or so -> If zinit then fails on a new/changed project, one could add another annotated input and run the test and then do fast iterations on the code to also catch that edge case.

The ghr test could then be reduced to a few reliable projects which actually exercise different code paths.
